nixos/tests/certmgr.nix

```diff
@@ -0,0 +1,125 @@
+            algo = "rsa"; size = 4096; };
+            names = [
+              {
+                C = "US";
+                L = "San Francisco";
+                O = "Internet Widgets, LLC";
+                OU = "Certificate Authority";
+                ST = "California";
+              }
+            ];
+        })} | ${pkgs.cfssl}/bin/cfssljson -bare ca
+      '';
+    };
+
+    services.nginx = {
+      enable = true;
+      virtualHosts = lib.mkMerge (map (host: {
+        ${host} = {
+          sslCertificate = "/tmp/${host}-cert.pem";
+          sslCertificateKey = "/tmp/${host}-key.pem";
+          extraConfig = ''
+            ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
+          '';
+          onlySSL = true;
+          serverName = host;
+          root = pkgs.writeTextDir "index.html" "It works!";
+        };
+      }) [ "imp.example.org" "decl.example.org" ]);
+    };
+
+    systemd.services.nginx.wantedBy = lib.mkForce [];
+
+    systemd.services.certmgr.after = [ "cfssl.service" ];
+    services.certmgr = {
+      enable = true;
+
+      specs =
+      let
+        spec = host: service: {
+          action = "restart";
+          authority = {
+            file = {
+              group = "nobody";
+              owner = "nobody";
+              path = "/tmp/${host}-ca.pem";
+            };
+            label = "www_ca";
+            profile = "three-month";
+            remote = "localhost:8888";
+          };
+          certificate = {
+            group = "nobody";
+            owner = "nobody";
+            path = "/tmp/${host}-cert.pem";
+          };
+          private_key = {
+            group = "nobody";
+            mode = "0600";
+            owner = "nobody";
+          path = "/tmp/${host}-key.pem";
+        };
+        request = {
+          CN = host;
+          hosts = [ host "www.${host}" ];
+          key = {
+            algo = "rsa";
+            size = 2048;
+          };
+          names = [
+            {
+              C = "US";
+              L = "San Francisco";
+              O = "Example, LLC";
+              ST = "CA";
+            }
+          ];
+        };
+        inherit service;
+      };
+    in
+    {
+      impl = toString (pkgs.writeText "test.json" (builtins.toJSON (spec "imp.example.org" "nginx")));
+      decl = spec "decl.example.org" "nginx";
+    };
+
+    };
+  };
+
+  testScript = ''
+    $machine->waitForUnit('cfssl.service');
+    $machine->waitUntilSucceeds('ls /tmp/decl.example.org-ca.pem');
+    $machine->waitUntilSucceeds('ls /tmp/decl.example.org-key.pem');
+    $machine->waitUntilSucceeds('ls /tmp/decl.example.org-cert.pem');
+    $machine->waitUntilSucceeds('ls /tmp/imp.example.org-ca.pem');
+    $machine->waitUntilSucceeds('ls /tmp/imp.example.org-key.pem');
+    $machine->waitUntilSucceeds('ls /tmp/imp.example.org-cert.pem');
+    $machine->waitForUnit('nginx.service');
+    $machine->succeed('[ "1" -lt "$(journalctl -u nginx | grep "Starting Nginx" | wc -l)" ]');
```

Comment:
Why this weird check? Maybe a `$machine->sleep(3)` will work instead if the `waitForUnit` really isn't enough.

Reply:
The check is there to confirm that nginx is in fact restarted by certmgr after certificate issue. `Starting nginx ...` would be present in the journal more than once if nginx was restarted. `waitForUnit` will terminate just fine even if certmgr restart action is set to no-op/dummy.

Reply:
Ah I see